File: vripper-server/src/main/java/me/vripper/host/TurboImageHost.java
package me.vripper.host;

import lombok.extern.slf4j.Slf4j;
import me.vripper.exception.HostException;
import me.vripper.exception.XpathException;
import me.vripper.jpa.domain.Image;
import me.vripper.services.HostService;
import me.vripper.services.XpathService;
import org.apache.http.client.protocol.HttpClientContext;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.w3c.dom.Document;
import org.w3c.dom.Node;

@Service
@Slf4j
public class TurboImageHost extends Host {

  private static final String host = "turboimagehost.com";
  private static final String TITLE_XPATH = "//div[contains(@class,'titleFullS')]/h1";
  private static final String IMG_XPATH = "//img[@id='imageid']";

  private final HostService hostService;
  private final XpathService xpathService;

  @Autowired
  public TurboImageHost(HostService hostService, XpathService xpathService) {
    this.hostService = hostService;
    this.xpathService = xpathService;
  }

  @Override
  public String getHost() {
    return host;
  }

  @Override
  public String getLookup() {
    return host;
  }

  @Override
  public HostService.NameUrl getNameAndUrl(final Image image, final HttpClientContext context)
      throws HostException {

    Document doc = hostService.getResponse(image.getUrl(), context).getDocument();

    String title;
    try {
      log.debug(
          String.format("Looking for xpath expression %s in %s", TITLE_XPATH, image.getUrl()));
      Node titleNode = xpathService.getAsNode(doc, TITLE_XPATH);
      log.debug(String.format("Resolving name for %s", image.getUrl()));
      if (titleNode != null) {
        title = titleNode.getTextContent().trim();
      } else {
        title = null;
      }
    } catch (XpathException e) {
      throw new HostException(e);
    }

    if (title == null || title.isEmpty()) {
      title = hostService.getDefaultImageName(image.getUrl());
    }

    try {
      Node urlNode = xpathService.getAsNode(doc, IMG_XPATH);
      return new HostService.NameUrl(
          title, urlNode.getAttributes().getNamedItem("src").getTextContent().trim());
    } catch (Exception e) {
      throw new HostException("Unexpected error occurred", e);
    }
  }
}

File: vripper-core/src/main/kotlin/me/vripper/host/TurboImageHost.kt
package me.vripper.host

import me.vripper.exception.HostException
import me.vripper.exception.XpathException
import me.vripper.services.download.ImageDownloadRunnable
import me.vripper.utilities.LoggerDelegate
import me.vripper.utilities.XpathUtils
import org.w3c.dom.Node

internal class TurboImageHost : Host("turboimagehost.com", 14) {
    private val log by LoggerDelegate()

    @Throws(HostException::class)
    override fun resolve(
        context: ImageDownloadRunnable.Context
    ): Pair<String, String> {
        val document = fetchDocument(context.imageEntity.url, context)
        var title: String?
        title = try {
            log.debug(String.format("Looking for xpath expression %s in %s", TITLE_XPATH, context.imageEntity.url))
            val titleNode: Node? = XpathUtils.getAsNode(document, TITLE_XPATH)
            log.debug(String.format("Resolving name for %s", context.imageEntity.url))
            titleNode?.textContent?.trim { it <= ' ' }
        } catch (e: XpathException) {
            throw HostException(e)
        }
        if (title.isNullOrEmpty()) {
            title = getDefaultImageName(context.imageEntity.url)
        }
        val urlNode: Node = XpathUtils.getAsNode(document, IMG_XPATH)
            ?: throw HostException(
                String.format(
                    "Xpath '%s' cannot be found in '%s'",
                    IMG_XPATH,
                    context.imageEntity.url
                )
            )
        return Pair(title, urlNode.attributes.getNamedItem("src").textContent.trim { it <= ' ' })
    }

    companion object {
        private const val TITLE_XPATH = "//div[contains(@class,'titleFullS')]/h1"
        private const val IMG_XPATH = "//img[@id='imageid']"
    }
}